Feature/ITX2SKILLS-269/class_structure_organization

Oliterra opened the pull request:
1) The following classes have been modified:
- Command
- CommandDescriptor
- ConsoleInterpreter
- ConsoleLineParser
- CommandFactory
- CommandExecutor
- CommandResult
- ConsoleCommandResultViewer

2) Added DTO to get the result of each command execution method and mappers for them

3) Each command was added as a class

4) Command exception handler was added

5) Command params validator was added

Oliterra committed:
TX2SKILLS-269 Modification of the command execution mechanism

The following classes have been modified:
Command
CommandDescriptor
ConsoleInterpreter
ConsoleLineParser
CommandFactory
CommandExecutor
CommandResult
ConsoleCommandResultViewer

Added DTO to get the result of each command execution method and mappers for them

Each command was added as a class

CommandExceptionHandler was added

CommandParamsValidator was added

Oliterra committed:
TX2SKILLS-269 Refactoring

Unnecessary logger settings file, methods, exceptions and variables have been deleted

smuschevich commented on main/src/main/java/edu/bank/Application.java:
```diff
@@ -10,7 +10,7 @@ public class Application {
         ConfigurableApplicationContext context = SpringApplication.run(Application.class, args);
-        ConsoleInterpreterImpl consoleInterpreter = context.getBean(ConsoleInterpreterImpl.class);
+        ConsoleInterpreter consoleInterpreter = context.getBean(ConsoleInterpreter.class);
         consoleInterpreter.interpretConsoleInput();
```
Introduce ConsoleInterpreterRunner to start ConsoleInterpreter

smuschevich commented on main/src/main/java/edu/bank/console/ConsoleInterpreter.java:
```diff
@@ -1,5 +1,41 @@
+    private static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+
+    public <T> void interpretConsoleInput() {
+        commandResultViewer.showWelcomeMessage();
```
Create separate command for showWelcomeMessage

smuschevich commented on main/src/main/java/edu/bank/console/ConsoleInterpreter.java:
```diff
@@ -1,5 +1,41 @@
+        while (true) {
+            try {
+                String consoleInput = reader.readLine();
+                if (commandResultViewer.isTheCycleInterrupted(consoleInput)) break;
```
Created separate command for exit

smuschevich commented on main/src/main/java/edu/bank/model/command/CommandDescription.java:
```diff
@@ -0,0 +1,12 @@
+@Data
+public class CommandDescription {
+
+    protected String name;
```
make name private

smuschevich commented on main/src/main/java/edu/bank/console/ConsoleInterpreter.java:
```diff
@@ -1,5 +1,41 @@
+    private final CommandExceptionHandler commandExceptionHandler;
+    private static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+
+    public <T> void interpretConsoleInput() {
```
Remove generic from method level

smuschevich commented on main/src/main/java/edu/bank/console/ConsoleInterpreter.java:
```diff
@@ -1,5 +1,41 @@
+                CommandResult<T> commandResult = commandExecutor.executeCommand(commandDescription);
+                commandResultViewer.showResult(commandResult);
+            } catch (Exception e) {
+                CommandResult<ErrorDTO> commandResult = commandExceptionHandler.handleException(e.getCause());
```
Use original exception as well.

smuschevich commented on main/src/main/java/edu/bank/result/CommandResultMapper.java:
```diff
@@ -0,0 +1,6 @@
+
+public interface CommandResultMapper<T> {
+
+    void showInfo(T t);
```
`showInfo` should just map object to string, and do not print to the console

smuschevich commented on main/src/main/java/edu/bank/console/ConsoleCommandResultViewer.java:
```diff
@@ -1,11 +1,42 @@
+    public <T> void showResult(CommandResult<T> commandResult) {
+        T result = commandResult.getResult();
+        CommandResultMapper<T> mapper = commandResult.getCommandResultMapper();
+        if (result == null) printSuccessMessage();
```
Always put logic code if-statement into block (curly braces)

smuschevich replied:
Avoid nullable result in CommandResult

smuschevich commented on main/src/main/java/edu/bank/console/ConsoleCommandResultViewer.java:
```diff
@@ -1,11 +1,42 @@
+        if (result == null) printSuccessMessage();
+        else if (mapper != null) mapper.showInfo(result);
+        else if (result.getClass().equals(ErrorDTO.class)) printErrorMessage(((ErrorDTO) result).getMessage());
+        else System.out.println(result);
```
replace `System.out.println` everywhere by slf4j with logback

smuschevich replied:
Use custom console pattern just for your package, other libs packages (e.g. Spring) should use default one.

smuschevich commented on main/src/main/java/edu/bank/command/CommandExecutor.java:
```diff
@@ -1,7 +1,27 @@
+
+    @SneakyThrows
+    public <T, E> CommandResult<E> executeCommand(CommandDescription commandDescription) throws IOException {
+        CommandInfo commandInfo = commandFactory.getCommandInfoByCommandName(commandDescription.getName());
```
All logic of command creation should be encapsulated inside the factory

Oliterra committed:
Correction of comments

Java version was switched to 17

The properties was applied in maven pom (java.version)

Spring boot maven plugin was applied

Docker-compose was applied

ConsoleInterpreterRunner was introduced to init ConsoleInterpreter main method

Separate commands for showing welcome message and exit from application were created

Changed field access modifier in console interpreter class

Generic was removed from method level in console interpreter class

Mappers were modified to just map objects to string, and do not print them to the console

All if-else constructions logic code were placed into blocks(curly braces)

All void returning services  methods were changed to return string(some message about their execution). Thus, there is no longer a nullable result in Command Result.

The exception handler has been modified to log the full exception chain and show the user only business logic exception messages and invalid input messages.

System.out.println was replaced everywhere by slf4j with logback

All logic of command creation was encapsulated inside the command factory class

Oliterra committed:
ITX2SKILLS-269 Correction of comments

Java version was switched to 17

The properties was applied in maven pom (java.version)

Spring boot maven plugin was applied

Docker-compose was applied

ConsoleInterpreterRunner was introduced to init ConsoleInterpreter main method

Separate commands for showing welcome message and exit from application were created

Changed field access modifier in console interpreter class

Generic was removed from method level in console interpreter class

Mappers were modified to just map objects to string, and do not print them to the console

All if-else constructions logic code were placed into blocks(curly braces)

All void returning services  methods were changed to return string(some message about their execution). Thus, there is no longer a nullable result in Command Result.

The exception handler has been modified to log the full exception chain and show the user only business logic exception messages and invalid input messages.

System.out.println was replaced everywhere by slf4j with logback

All logic of command creation was encapsulated inside the command factory class

Oliterra committed:
Merge branch 'feature/ITX2SKILLS-269/class_structure_organization' of https://github.com/NadezhdaTarasova/BankSimulator into feature/ITX2SKILLS-269/class_structure_organization

Oliterra committed:
Dockerfile was updated

image in dockerfile was changed to match the java version

smuschevich commented:
- BusinessLogicException will extend Exception
- introduce ValidationException with parent BusinessLogicException

smuschevich commented on main/src/main/java/edu/bank/console/ConsoleInterpreter.java:
```diff
@@ -1,5 +1,48 @@
+                }
+                CommandResult<?> commandResult = commandExecutor.executeCommand(commandDescription);
+                commandResultViewer.showResult(commandResult);
+            } catch (ExitRequest e) {
```
Exception vs flag in CommandResult

smuschevich commented on main/src/main/java/edu/bank/model/command/Command.java:
```diff
@@ -0,0 +1,12 @@
+
+    void validateCommand(CommandDescription commandDescription) throws IOException;
+
+    <T> CommandResult<T> executeCommand(CommandDescription commandDescription) throws IOException, ReflectiveOperationException;
```
Move generic to class level

smuschevich commented on main/src/main/java/edu/bank/command/CommandExecutor.java:
```diff
@@ -1,7 +1,22 @@
+    private final CommandFactory commandFactory;
+
+    @SneakyThrows
+    public <T> CommandResult<T> executeCommand(CommandDescription commandDescription) {
```
Add to signature declaration of business exception.

smuschevich commented on main/src/main/java/edu/bank/command/CommandExecutor.java:
```diff
@@ -1,7 +1,22 @@
+
+    private final CommandFactory commandFactory;
+
+    @SneakyThrows
```
Investigate and apply @SneakyThrows only if necessary.

Oliterra committed:
ITX2SKILLS-269 Correction of comments

1) In Command interface generic was moved to class level
2) Param “command execution services package name” was moved to commands-list.yml file(like “commands command package name”)
3) BusinessLogicException now extends Exception
4) A ValidationException exception was created(extends the BusinessLogicException exception)
5) For cycle in entry point “interpretConsoleInput”(im ConsoleInterpreter class) was simplified by boolean counter(instead of integer counter). “Magic number”(if i == 0) was removed.
6) The basic implementation of the “ExecuteCommand” method was created in the BaseCommand class.
7) Added file logback.xml with 3 logger patterns (custom console pattern just for edu.bank package, detailed console pattern for spring packages and custom pattern for internal errors(ExceptionHandler class))
8) All versions of dependencies in maven .pom are now available via variables
9) Docker directory was moved to root folder
10) Dockerfile was moved to root folder
11) Dockerfile was modified (proper version was used for openjdk, spring boot maven plugin was applied)
12) The structure of the project packages has been changed to a more logical one
13) Installation instruction was added to README.md file

Oliterra committed:
Update README.md

Oliterra committed:
Update README.md

Oliterra committed:
Dockerfile was updated

Was applied a multi-staged build(building stage image - maven:3.8.5-openjdk-17-slim, second stage image - openjdk:17-alpine)

Oliterra committed:
Update README.md
